File: FindMinimumInRotatedSortedArray.java
public class FindMinimumInRotatedSortedArray {
    // public int findMin(int[] nums) {
    //     int start = 0;
    //     int end = nums.length - 1;
        
    //     if (nums.length <= 1) {
    //         return nums[0];
    //     }
        
    //     if (nums[start] < nums[end]) {
    //         return nums[start];
    //     }
        
    //     while (start < end) {
    //         start ++;
    //         end --;
    //         if (nums[start] < nums[start - 1]) {
    //             return nums[start];
    //         }
    //         if (nums[end] > nums[end + 1]) {
    //             return nums[end + 1];
    //         }
    //     }
        
    //     return 0;
    // }

    public int findMin(int[] nums) {
        int start = 0;
        int end = nums.length - 1;
        
        if (nums.length <= 1) {
            return nums[0];
        }
        
        if (nums[start] < nums[end]) {
            return nums[start];
        }
        
        while (start < end) {
            if (start + 1 == end) {
                return Math.min(nums[start], nums[end]);
            }
            int middle = (start + end)/2;
            if (nums[start] < nums[end]) {
                return nums[start];
            } else {
                if (nums[start] > nums[middle]) {
                    end = middle;
                } else {
                    start = middle;
                } 
            }
        }
        
        return 0;
    }
}

File: FindMinimumInRotatedSortedArray.py
class FindMinimumInRotatedSortedArray(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return min(nums[0], nums[1])
        
        left = 0
        right = len(nums) - 1
        if nums[left] < nums[right]:
            return nums[left]
        
        result = nums[0]
        while left < right:
            if nums[left] < nums[right]:
                return min(nums[left], result)
            result = min(nums[right], result)
            mid = int(math.floor(right + left)/2)
            result = min(nums[mid], result)
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid - 1
        
        return result